### PR TITLE
Validation unit test

### PR DESCRIPTION
**Pull request by yangcao77**
This PR is for #47 
Added unittests for validation code for all types of projects. 

```
  PFE - unit test
    actions module
      combinational testing of build function
        ✓ combo1 => args: {}
        ✓ combo2 => args: {"projectID":"testprojectID"}
      combinational testing of restart function
        ✓ combo1 => args: {}
        ✓ combo2 => args: {"projectID":"testprojectID"}
        ✓ combo3 => args: {"projectID":"testprojectID","startMode":"invalidStartMode"}
        ✓ combo4 => args: {"projectID":"testprojectID","startMode":"run"}
        ✓ combo5 => args: {"projectID":"testprojectID","startMode":"run"}
      combinational testing of enableautobuild & disableautobuild function
        ✓ combo1enableautobuild => args: {}
        ✓ combo1disableautobuild => args: {}
        ✓ combo2enableautobuild => args: {"projectID":"dummynodeproject"}
        ✓ combo2disableautobuild => args: {"projectID":"dummynodeproject"}
        ✓ combo3enableautobuild => args: {"projectID":"dummyswiftproject"}
        ✓ combo3disableautobuild => args: {"projectID":"dummyswiftproject"}
        ✓ combo4enableautobuild => args: {"projectID":"dummyspringproject"}
        ✓ combo4disableautobuild => args: {"projectID":"dummyspringproject"}
        ✓ combo5enableautobuild => args: {"projectID":"dummymicroprofileproject"}
        ✓ combo5disableautobuild => args: {"projectID":"dummymicroprofileproject"}
        ✓ combo6enableautobuild => args: {"projectID":"dummygoproject"}
        ✓ combo6disableautobuild => args: {"projectID":"dummygoproject"}
        ✓ combo7enableautobuild => args: {"projectID":"dummyappsodyproject"}
        ✓ combo7disableautobuild => args: {"projectID":"dummyappsodyproject"}
      combinational testing of validate function
        ✓ combo1 => validate: request missing project type and location
        ✓ combo2 => validate: request with invalidProjectType
        ✓ combo3 => validate: request with non-exist project location
        ✓ combo4 => validate: good microprofile project
        ✓ combo5 => validate: microprofile project with missing Dockerfile
        ✓ combo6 => validate: microprofile project with missing server.xml
        ✓ combo7 => validate: microprofile project with root user dockerfile-build
        ✓ combo8 => validate: microprofile project with bad inner most pom.xml elements
        ✓ combo9 => validate: microprofile project with bad pom.xml due to missing liberty maven plugin
        ✓ combo10 => validate: microprofile project with bad pom.xml due to bad profile
        ✓ combo11 => validate: microprofile project with bad pom.xml syntax
        ✓ combo12 => validate: good spring project
        ✓ combo13 => validate: spring project with missing Dockerfile
        ✓ combo14 => validate: spring project with bad inner most pom.xml elements
        ✓ combo15 => validate: good nodejs project
        ✓ combo16 => validate: nodejs project with missing Dockerfile
        ✓ combo17 => validate: nodejs project with missing package.json
        ✓ combo18 => validate: good swift project
        ✓ combo19 => validate: swift project with missing Dockerfile-tools
        ✓ combo20 => validate: swift project with missing Package.swift
        ✓ combo21 => validate: good python project
        ✓ combo22 => validate: python project with missing Dockerfile
        ✓ combo23 => validate: good appsody project
        ✓ combo24 => validate: appsody project with missing requiredFiles


  45 passing (302ms)
```

### CODE REVIEW

**Comment by yangcao77** [comment]
@maysunfaisal @rajivnathan Can you help review this PR? Thanks.

File: src/pfe/file-watcher/server/src/projects/LibertyValidator.ts [hunk 1]
```diff
@@ -143,10 +143,12 @@ export class LibertyValidator extends Validator {
                                 let libertyPluginFound = false;
                                 try {
                                     if (mcprofile.build[0].plugins[0].plugin) {
-                                        mcprofile.build[0].plugins[0].plugin.forEach(async (_plugin: any) => {
+                                         for ( const _plugin of  mcprofile.build[0].plugins[0].plugin ) {
+                                            logger.logInfo("_plugin.groupId: " + _plugin.groupId[0]);
```
Comment (maysunfaisal) [inline src/pfe/file-watcher/server/src/projects/LibertyValidator.ts]: think we should have these logs as logTrace or logDebug

File: src/pfe/file-watcher/server/src/projects/LibertyValidator.ts [hunk 2]
```diff
@@ -143,10 +143,12 @@ export class LibertyValidator extends Validator {
...
+                                            logger.logInfo("_plugin.artifactId: " + _plugin.artifactId[0]);
                                             if (_plugin.groupId[0] === "net.wasdev.wlp.maven.plugins" && _plugin.artifactId[0] === "liberty-maven-plugin") {
                                                 libertyPluginFound = true;
-
+                                                logger.logInfo("_plugin.extensions: " + _plugin.extensions[0]);
```
Comment (maysunfaisal) [inline src/pfe/file-watcher/server/src/projects/LibertyValidator.ts]: same as above?

File: src/pfe/file-watcher/server/test/unit-test/tests/actions.module.test.ts [hunk 3]
```diff
@@ -206,4 +259,435 @@ export function actionsTestModule(): void {
...
+            },
+            "combo11": {
+                "description": "microprofile project with bad pom.xml syntax",
+                "args": {
+                    projectType: "liberty",
+                    projectID: "javaMicroProfileTemplate",
+                    location: path.join(process.env.CW_WORKSPACE, "javaMicroProfileTemplate")
+                },
+                "result": "failed",
+                "error": "POM parsing error"
```
Comment (maysunfaisal) [inline src/pfe/file-watcher/server/test/unit-test/tests/actions.module.test.ts]: I think you missed out on one Liberty validation: `libertyProject.validator.parentPOM.label` which is in the first catch block of `LibertyValidator.validatePOM`

Adding this probably means having another copy of a garbage pom i think? I shall leave this upto you

File: src/pfe/file-watcher/server/src/projects/LibertyValidator.ts (hunk 1 above)
Comment (yangcao77) [reply src/pfe/file-watcher/server/src/projects/LibertyValidator.ts]: That's the log message I used for debugging purpose. Should be removed. Thanks for catching it.

File: src/pfe/file-watcher/server/src/projects/LibertyValidator.ts (hunk 2 above)
Comment (yangcao77) [reply src/pfe/file-watcher/server/src/projects/LibertyValidator.ts]: updated

File: src/pfe/file-watcher/server/test/unit-test/tests/actions.module.test.ts (hunk 3 above)
Comment (yangcao77) [reply src/pfe/file-watcher/server/test/unit-test/tests/actions.module.test.ts]: Added.

File: src/pfe/file-watcher/server/package.json [hunk 4]
```diff
@@ -14,8 +14,8 @@
     "test": "npm run unit:test && npm run functional:test",
     "unit:setup": "./test/scripts/unit_setup.sh",
     "unit:cleanup": "./test/scripts/unit_cleanup.sh",
-    "unit:test": "npm run unit:setup ; NODE_ENV=test CW_LOCALES_DIR=$(pwd)/src/utils/locales/ CW_WORKSPACE=$(pwd) CW_PROJECTDATA_DIR=$(pwd)/test/resources/ nyc mocha test/unit-test/unit.test.ts ; npm run unit:cleanup",
-    "unit:test:xml": "npm run unit:setup ; NODE_ENV=test CW_LOCALES_DIR=$(pwd)/src/utils/locales/  CW_WORKSPACE=$(pwd) CW_PROJECTDATA_DIR=$(pwd)/test/resources/ nyc mocha test/unit-test/unit.test.ts --colors --reporter mocha-jenkins-reporter ; npm run unit:cleanup",
+    "unit:test": "npm run unit:setup ; NODE_ENV=test CW_LOCALES_DIR=$(pwd)/src/utils/locales/ CW_WORKSPACE=$(pwd)/test/resources CW_PROJECTDATA_DIR=$(pwd)/test/resources/ nyc mocha test/unit-test/unit.test.ts ; npm run unit:cleanup",
```
[the remark this answers is not in the document]
Comment (yangcao77) [reply src/pfe/file-watcher/server/package.json]: It's cleaned in the cleanup script.

File: src/pfe/file-watcher/server/src/projects/actions.ts [hunk 5]
```diff
@@ -51,7 +50,10 @@ const validate = async function(args: IProjectActionParams): Promise<{ operation
         "projectType": projectType,
         "location": location
     } as ProjectInfo;
-
+    const projectName = location.split("/").pop();
+    if (args.extensionID) {
+        projectInfo.extensionID = args.extensionID;
```
[the remark this answers is not in the document]
Comment (yangcao77) [reply src/pfe/file-watcher/server/src/projects/actions.ts]: it doesn't need to, it's still saved as `extensionID` in `projectInfo`

File: src/pfe/file-watcher/server/src/projects/libertyProject.ts [hunk 6]
```diff
@@ -218,7 +220,7 @@ export async function logBuildEvent(projectInfo: ProjectInfo, msg: String, isErr
 
     logger.logProjectInfo(`Writing to build log at ${buildLog} :\n\t${fullMsg}`, projectInfo.projectID);
 
-    fs.appendFile(buildLogPath, fullMsg, (err) => {
+    await fs.appendFile(buildLogPath, fullMsg, (err) => {
```
[the remark this answers is not in the document]
Comment (yangcao77) [reply src/pfe/file-watcher/server/src/projects/libertyProject.ts]: This does not have to be changed to await. I will change back

File: src/pfe/file-watcher/server/src/projects/actions.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (makandre) [reply src/pfe/file-watcher/server/src/projects/actions.ts]: So yes the `extensionID` here is fine.  However I think we may actually need the `projectType.projectID` as well.  This is probably missed in the following PR https://github.com/eclipse/codewind/pull/147 :(

[the remark this answers is not in the document]
Comment (makandre) [reply src/pfe/file-watcher/server/src/projects/actions.ts]: Discussed on chat and:

1) just need to make sure `args.extensionID` is a full path
2) need to pass in `projectID` and `language` as well

**Pull request merged**